lib/handler.js

```diff
@@ -50,9 +49,12 @@ function writeErrorToResponse(err, req, res, options) {
 
   options = options || {};
 
-  if (res._header) {
-    debug('Response was already sent, closing the underlying connection');
-    return req.socket.destroy();
+  // See https://nodejs.org/api/http.html#http_response_headerssent
+  if (res.headersSent) {
+    debug('Response was already sent. Skipping error response.');
+    // We should not destroy the request socket as it causes Error: write EPIPE
```

Comment:
For me, the EPIPE error is still happening without destroying the request socket. Do we want to throw a meaningful error here instead?

Reply:
No. The EPIPE error is reported from the client side as the socket is destroyed by the server.

Reply:
Okay, any idea on why it's failing for me on node 10?
```
  49 passing (236ms)
  1 failing

  1) strong-error-handler
       handles response headers already sent without destroying the request:
     Error: write EPIPE
      at WriteWrap.afterWrite [as oncomplete] (net.js:788:14)



npm ERR! Test failed.  See above for more details.
biniams-mbp:strong-error-handler badmike$ node -v
v10.11.0
biniams-mbp:strong-error-handler badmike$ npm -v
6.4.1
biniams-mbp:strong-error-handler badmike$ git br
* fix-write-error
  master
```

Reply:
A bit more verbose logs:
```
biniams-mbp:strong-error-handler badmike$ DEBUG=strong-error-handler npm t

> strong-error-handler@3.2.0 test /Users/badmike/loopback/strong-error-handler> mocha



  strong-error-handler
  strong-error-handler Initializing with options {"log":false} +0ms
  strong-error-handler Handling Error: an error
    at givenErrorHandlerForError (/Users/badmike/loopback/strong-error-handler/test/handler.test.js
:859:23)
    at Context.<anonymous> (/Users/badmike/loopback/strong-error-handler/test/handler.test.js:43:7)
    at callFnAsync (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runnable.js
:400:21)
    at Test.Runnable.run (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runna
ble.js:342:7)
    at Runner.runTest (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.j
s:455:10)
    at /Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:573:12
    at next (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:369:14)
    at /Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:379:7
    at next (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:303:14)
    at /Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:342:7
    at done (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runnable.js:319:5)
    at callFn (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runnable.js:395:7)
    at Hook.Runnable.run (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runnable.js:364:7)
    at next (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:317:10)
    at Immediate.<anonymous> (/Users/badmike/loopback/strong-error-handler/node_modules/mocha/lib/runner.js:347:5)
    at runCallback (timers.js:694:18)
    at tryOnImmediate (timers.js:665:5)
    at processImmediate (timers.js:647:5) +40ms
  strong-error-handler Response was already sent. Skipping error response. +1ms
    1) handles response headers already sent without destroying the request


  0 passing (71ms)
  1 failing

  1) strong-error-handler
       handles response headers already sent without destroying the request:
     Error: write EPIPE
      at WriteWrap.afterWrite [as oncomplete] (net.js:788:14)
```

Reply:
not sure why `givenErrorHandlerForError` generates an instance of `Error` if there isn't one passed into it.

Reply:
I see it now. It's interesting, `res.end` should half-close the socket and it should allow in-progress request to finish.